### PR TITLE
Fix HTTP caching

### LINKED ISSUE #17686: HTTP caching is not working
Currently HTTP caching is not working. Seems like a bug in the logic which adds responses to the cache. 
Problematic code block:https://github.com/ballerina-platform/ballerina-lang/blob/799db2e2caae022c32e21dd004975246eeb35932/stdlib/http/src/main/ballerina/src/http/caching/http_cache.bal#L53

### PR DESCRIPTION
**Pull request by pubudu91**
Fixes #17686

## Check List 
- [ ] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Required Balo version update
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

**Comment by chamil321** [comment]
Shall we add an integration test along this change?

**Comment by pubudu91** [comment]
@chamil321 Added one which tests the basics

File: tests/jballerina-integration-test/src/test/java/org/ballerinalang/test/service/http/sample/HTTPCachingTestCase.java [hunk 1]
```diff
@@ -0,0 +1,64 @@
+/*
+ *  Copyright (c) 2017, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (chamil321) [inline tests/jballerina-integration-test/src/test/java/org/ballerinalang/test/service/http/sample/HTTPCachingTestCase.java]:
```suggestion
 *  Copyright (c) 2019, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```

File: stdlib/http/src/main/ballerina/src/http/caching/http_cache.bal [hunk 2]
```diff
@@ -40,61 +39,86 @@ public type HttpCache object {
...
+        ResponseCacheControl? respCC = inboundResp.cacheControl;
+        boolean allowedByCacheControl = false;
+
+        if (respCC is ResponseCacheControl) {
+            if (respCC.maxAge >= 0 || (self.isShared && (respCC.sMaxAge >= 0)) || !respCC.isPrivate) {
+                allowedByCacheControl = true;
             }
         }
+
+        if (allowedByCacheControl || inboundResp.hasHeader(EXPIRES) || isCacheableStatusCode(inboundResp.statusCode)) {
```
Comment (chamil321) [inline stdlib/http/src/main/ballerina/src/http/caching/http_cache.bal]:
```suggestion
        return (allowedByCacheControl || inboundResp.hasHeader(EXPIRES) || isCacheableStatusCode(inboundResp.statusCode));
```

**Pull request merged**